atc/engine/builder/step_factory.go

```diff
@@ -124,8 +124,18 @@ func (factory *stepFactory) TaskStep(
 	containerMetadata db.ContainerMetadata,
 	delegate exec.TaskDelegate,
 ) exec.Step {
-	sum := sha1.Sum([]byte(plan.Task.Name))
-	containerMetadata.WorkingDirectory = filepath.Join("/tmp", "build", fmt.Sprintf("%x", sum[:4]))
+
+	var wd string
+	configRunDir := plan.Task.Config.Run.Dir
```

Comment:
This will only work if the task has its config statically provided like this:

```yaml
task: foo
config: # ...
```

For this to also work with `file:` you'll need to modify `task_step.go` to determine this after the config is loaded instead.